### ISSUE TITLE
[REVIEW]: multivar_horner: A Python package for computing Horner factorisations of multivariate polynomials

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2392]
**Submitting author:** @MrMinimal64 (<a href="http://orcid.org/0000-0002-1819-6975">Jannik Michelfeit</a>)
**Repository:** <a href="https://github.com/MrMinimal64/multivar_horner" target ="_blank">https://github.com/MrMinimal64/multivar_horner</a>
**Version:** 2.1.1
**Editor:** @dpsanders
**Reviewers:** @henrik227, @saschatimme
**Archive:** <a href="https://doi.org/10.5281/zenodo.4062422" target="_blank">10.5281/zenodo.4062422</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/0b514c6894780f3cc81ed88c141631d4/status.svg)](https://joss.theoj.org/papers/0b514c6894780f3cc81ed88c141631d4)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/0b514c6894780f3cc81ed88c141631d4"><img src="https://joss.theoj.org/papers/0b514c6894780f3cc81ed88c141631d4/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/0b514c6894780f3cc81ed88c141631d4/status.svg)](https://joss.theoj.org/papers/0b514c6894780f3cc81ed88c141631d4)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/MrMinimal64/multivar_horner" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@henrik227, @alex-konovalov and @saschatimme, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @dpsanders know.

✨ **Please try and complete your review in the next six weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/MrMinimal64/multivar_horner">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@MrMinimal64) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

---
### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/MrMinimal64/multivar_horner">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@MrMinimal64) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @henrik227, @alex-konovalov it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1109/MCSE.2011.37 is OK
- 10.1145/2833157.2833162 is OK

MISSING DOIs

- https://doi.org/10.1098/rstl.1819.0023 may be missing for title: XXI. A new method of solving numerical equations of all orders, by continuous approximation
- https://doi.org/10.1201/9781315120195-1 may be missing for title: Introduction to numerical analysis
- https://doi.org/10.1007/978-3-030-01177-2_5 may be missing for title: Fast Interpolation and Fourier Transform in High-Dimensional Spaces
- https://doi.org/10.1145/980175.980179 may be missing for title: Greedy algorithms for optimizing multivariate Horner schemes
- https://doi.org/10.1007/s006070070002 may be missing for title: On the multivariate Horner scheme II: running error analysis
- https://doi.org/10.15807/jorsj.51.29 may be missing for title: Efficient evaluation of polynomials and their partial derivatives in homotopy continuation methods
- https://doi.org/10.1090/s0025-5718-1990-0993925-1 may be missing for title: Evaluation of multivariate polynomials and their derivatives
- https://doi.org/10.1016/j.cpc.2013.05.008 may be missing for title: Improving multivariate Horner schemes with Monte Carlo tree search
- https://doi.org/10.1007/978-3-642-15582-6_55 may be missing for title: Efficient evaluation of large polynomials

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf)

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf)

**Comment by MrMinimal64** [comment]
@whedon set v2.1.0 as version

**Comment by whedon** [comment]
I'm sorry @MrMinimal64, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by MrMinimal64** [comment]
@dpsanders please bump the sofware version to v2.1.0 (latest release). Some useful API changes/clarifications have been included: https://github.com/MrMinimal64/multivar_horner/blob/master/CHANGELOG.rst#210-2020-06-15

**Comment by dpsanders** [comment]
@MrMinimal64 I don't think we need to do anything about the versioning until the review is complete.

**Comment by henrik227** [comment]
I am concerned that multivar_horner does not qualify as a `substantial scholarly effort' as described in the [review criteria](https://joss.readthedocs.io/en/latest/review_criteria.html#substantial-scholarly-effort), for the following reasons:

- The project is not very large (see the statistics [here](https://github.com/openjournals/joss-reviews/issues/2224#issuecomment-632872231)).
- It seems to be a rather straight-forward implementation of an existing algorithm (described in Ceberio & Kreinovich, 2004).

However, I do believe that multivar_horner is useful for anyone who needs to evaluate large multivariate polynomials.

Initially, I thought that the search-based approach mention in the `further reading' section could be a major contribution of multivar_horner. However, according to the [documentation](https://multivar-horner.readthedocs.io/en/latest/3_optimal_factorisation.html#optimal-horner-factorisations), this approach does not seem to have any practical benefits (yet?).

**Comment by MrMinimal64** [comment]
hello Henrik,

I do respect your opinion, but I would like to briefly argue against it in the following.

> The project is not very large (see the statistics here).

How large a project needs to be in order to fulfil the JOSS review criteria is open for discussion.
What can be said definitively however, is that many projects with a similar or even smaller footprint are being considered: 
https://github.com/openjournals/joss-reviews/issues/2428#issuecomment-653255875
https://github.com/openjournals/joss-reviews/issues/2438#issuecomment-653794412

> It seems to be a rather straight-forward implementation of an existing algorithm (described in Ceberio & Kreinovich, 2004).

First of all I would like to point out that ``multivar_horner`` contains many computational aspects not included in the solely theoretical outline of the algorithm in (Ceberio & Kreinovich, 2004).
The mathematical considerations only comprise the factorisation process.
How a found Horner factorisation can be represented and evaluated in a computationally efficient way is not within the scope of the literature.
Whether or not an implementation is “straight-forward”, in the end always remains highly subjective and even if it was, I think it could very well still qualify as “substantial scholarly effort”.

``multivar_horner`` can be considered as a substantial scholarly effort due to the following reasons:

- It is the first publicly available implementation of any kind of Horner factorisations for multivariate polynomials. Without it, the, as you call them, ``existing algorithms`` were existing actually only in theory or private and in absence of any benchmarks, tests and documentation.
- Because of their simple differentiation and integration, as well as their pleasant vector space structure, polynomials in multiple variables are a standard choice as interpolants and are fundamental in ordinary differential equation (ODE) and partial differential equation (PDE) solvers (see for example [1-4]).
- Horner factorisations have beneficial properties for handling polynomials as I am proving in the paper under review here
- How well all the theoretical approaches mentioned above can be “substantial” to academia depends on transferring them into practice by providing numerical availablility and applicability. That is exactly what this package does, which in turn makes it “substantial” to academia as well.
- That the package has been downloaded 11k times in total and on average 27 times a day in the last month (s. https://pepy.tech/project/multivar-horner) demonstrates that there is a demand for the project and its code quality is adequate

As an open source package helpful for a large part of the scientific community ``multivar_horner`` should hence be considered for publication in the JOSS.


[1] Floater, Michael S., and Kai Hormann. "Barycentric rational interpolation with no poles and high rates of approximation." Numerische Mathematik 107.2 (2007): 315-331.
[2] Cools, Ronald. "Constructing cubature formulae: the science behind the art." Acta numerica 6 (1997): 1-54.
[3] Temlyakov, Vladimir N. "Cubature formulas, discrepancy, and nonlinear approximation." Journal of Complexity 19.3 (2003): 352-391.
[4] Cools, Ronald. "Advances in multidimensional integration." Journal of computational and applied mathematics 149.1 (2002): 1-12.

**Comment by henrik227** [comment]
To clarify: I do not question that the package is useful. However, I cannot say for sure that this package "represent[s] not less than three months of work for an individual", and from the review criteria it is not clear to me how the usefulness of the software is weighed against the effort that went into it. I am well aware that it is difficult to objectively quantify the development effort of software.

I do not have a strong opinion on this; I just want to make sure that the package meets the standards of JOSS. If @dpsanders and @alex-konovalov consider the package a "substantial scholarly effort", I am ok with that.

**Comment by dpsanders** [comment]
@henrik227 Thanks for expressing your concern. 

I am waiting for @alex-konovalov's input, and for a further opinion which I hope to get next week.

**Comment by dpsanders** [comment]
@whedon add saschatimme as reviewer

**Comment by whedon** [comment]
OK, saschatimme is now a reviewer

**Comment by dpsanders** [comment]
@whedon add @saschatimme as reviewer

**Comment by whedon** [comment]
OK, @saschatimme is now a reviewer

**Comment by dpsanders** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf)

**Comment by MrMinimal64** [comment]
Due to the uncertain outcome of this review, I would like to submit this work to arXiv.org
Please consider endorsing me for this: https://arxiv.org/auth/endorse?x=9QAMJT

**Comment by MrMinimal64** [comment]
in the meantime a preprint version has been published on [arXiv.org](http://arxiv.org/abs/2007.13152)

**Comment by MrMinimal64** [comment]
The current paper version now contains some additional aspects:
- an abstract also used to the arXiv paper version
- mentioning of ``Maple`` as a software already supporting multivariate Horner factorisations
- reference to ``Mathematica`` offering 1D Horner factorisations
- mention the ability to a posteriori change the coefficients of a factorised polynomial (the ``multivar_horner`` factorisation process is coefficient agnostic)

@saschatimme, @alex-konovalov,  what is the status of the review?

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf)

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf)

**Comment by MrMinimal64** [comment]
Hello @dpsanders, @saschatimme, @alex-konovalov, @henrik227,
This review already started 2 month ago. 
What is the current status and the timeline for the next steps?
With kind regards,
Jannik Michelfeit

**Comment by dpsanders** [comment]
👋 @MrMinimal64, apologies for my absence.

👋 @henrik227, @alex-konovalov, @saschatimme: How are your reviews of this submission coming along? Hopefully we can have a first round of comments finished, say, by the end of August? Thanks!

**Comment by henrik227** [comment]
@dpsanders I was waiting on a decision regarding whether this project is considered a "substantial scholarly effort". Wouldn't it make sense to decide on that first?

**Comment by dpsanders** [comment]
@henrik227 That's a fair point. In my opinion this submission **does** meet the criteria for a "substantial scholarly effort" and should be reviewed.

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf)

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf)

**Comment by MrMinimal64** [comment]
I tried to incorporate all the objections made by the reviewers in the latest version of the paper.

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf)

**Comment by dpsanders** [comment]
👋 @alex-konovalov: Will you be able to complete your review of this paper soon?

**Comment by dpsanders** [comment]
👋 @saschatimme and @henrik227: Many thanks for your very useful feedback in the issues on the repo! Are you happy with the current version?

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: :point_left:

**Comment by henrik227** [comment]
> 👋 @saschatimme and @henrik227: Many thanks for your very useful feedback in the issues on the repo! Are you happy with the current version?

I still disagree with how operations are counted in this package, but I think it's ok since it's explained both in the paper and the documentation.

**Comment by saschatimme** [comment]
I am in the same boat as @henrik227 that I disagree with the way operations are counted in the package. The article states 
> multivar_horner implements a multivariate Horner scheme using the greedy heuristic presented in (Ceberio & Kreinovich, 2004). 

However, looking into  (Ceberio & Kreinovich, 2004) the way operations are counted there is actually how @henrik227 and I suggest they are supposed to be counted (to compute `x^3` takes 2 operations) and not as the author implemented it and describes it in the article where `x^3` only requires single operation.

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: :point_left:

**Comment by MrMinimal64** [comment]
The operations are now being counted as you suggest (solely the number of multiplications).
The optimal search heuristic now minimises the new operation count and finds "optimal" factorisations in this sense.
I also evaluate the relative numerical error in the benchmarks.
The paper as well as the documentation have been updated accordingly.

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: :point_left:

**Comment by MrMinimal64** [comment]
I updated the placement of the figures (now the paper has a length of 8 pages again)

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: :point_left:

**Comment by MrMinimal64** [comment]
@saschatimme, can you please check off the remaining items from your review checklist?
I guess @alex-konovalov is busy at the moment and won't finish his review.
@dpsanders, are there any remaining issues or objections?

**Comment by dpsanders** [comment]
I don't see a discussion in the paper about how the number of operations is counted.

@saschatimme and @henrik227, are you happy with the current version?

@alex-konovalov: Are you able to complete your review? If not, I will remove you as a reviewer.

**Comment by MrMinimal64** [comment]
> I don't see a discussion in the paper about how the number of operations is counted.

it started here:
https://github.com/MrMinimal64/multivar_horner/issues/17#issuecomment-681935918
Now the method of counting the operations is as the reviewers have suggested.

**Comment by dpsanders** [comment]
Yes thanks; I would like the paper to describe the implemented choice for counting operations. (Maybe it's already there but I didn't see it.)

**Comment by henrik227** [comment]
> Yes thanks; I would like the paper to describe the implemented choice for counting operations. (Maybe it's already there but I didn't see it.)

It's in the middle of page 2.

> @henrik227, are you happy with the current version?

I have some minor comments:

- Is it intentional that in the example right above the description of how operations are counted, the "old" way of counting operations is used? I'm referring to "27 operations (exponentiations, multiplications and additions)"
- "the number of additions [...] remains constant irrespective of the polynomial factorisation." To me, it's not obvious why this is true. Would it be possible to give some intuition or provide a reference?
- "The number of multiplications contained in the canonical representation of a polynomial consequently is equal to the sum of its exponents." Same as above. However, I don't think it's very important, so perhaps I would just remove it.

**Comment by MrMinimal64** [comment]
> I have some minor comments:

I tried to address all of your remarks. Please check if your concerns persist.

**Comment by MrMinimal64** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: :point_left:

**Comment by henrik227** [comment]
> I tried to address all of your remarks. Please check if your concerns persist.

Looks good!

**Comment by saschatimme** [comment]
@MrMinimal64  @dpsanders  This now also looks good for me!

**Comment by MrMinimal64** [comment]
Thanks for the fair review!

@dpsanders are there any open points?

**Comment by dpsanders** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Re-invite a reviewer (if they can't update checklists)
@whedon re-invite @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# All commands can be run on a non-default branch, to do this pass a custom 
# branch name by following the command with `from branch custom-branch-name`.
# For example:

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a dry run of accepting the paper and depositing with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by dpsanders** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02392/joss.02392/10.21105.joss.02392.pdf) :page_facing_up: :point_left:

**Comment by dpsanders** [comment]
It looks good to me!

Could you please create a software archive, e.g. at zenodo.org, and paste the version number and the DOI here? Thanks!

**Comment by dpsanders** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/MCSE.2011.37 is OK
- 10.1145/2833157.2833162 is OK

MISSING DOIs

- 10.1098/rstl.1819.0023 may be a valid DOI for title: XXI. A new method of solving numerical equations of all orders, by continuous approximation
- 10.1201/9781315120195-1 may be a valid DOI for title: Introduction to numerical analysis
- 10.1007/978-3-030-01177-2_5 may be a valid DOI for title: Fast Interpolation and Fourier Transform in High-Dimensional Spaces
- 10.1145/980175.980179 may be a valid DOI for title: Greedy algorithms for optimizing multivariate Horner schemes
- 10.1007/s006070070002 may be a valid DOI for title: On the multivariate Horner scheme II: running error analysis
- 10.15807/jorsj.51.29 may be a valid DOI for title: Efficient evaluation of polynomials and their partial derivatives in homotopy continuation methods
- 10.1090/s0025-5718-1990-0993925-1 may be a valid DOI for title: Evaluation of multivariate polynomials and their derivatives
- 10.1016/j.cpc.2013.05.008 may be a valid DOI for title: Improving multivariate Horner schemes with Monte Carlo tree search
- 10.1007/978-3-642-15582-6_55 may be a valid DOI for title: Efficient evaluation of large polynomials
- 10.1016/s0377-0427(02)00517-4 may be a valid DOI for title: Advances in multidimensional integration

INVALID DOIs

- None
```

**Comment by MrMinimal64** [comment]
@whedon check references

**Comment by whedon** [comment]
Checking the BibTeX entries failed with the following error: 

 Failed to parse BibTeX on value "doi" (NAME) [#<BibTeX::Bibliography data=[17]>, "@", #<BibTeX::Entry >, {:title=>["Efficient evaluation of large polynomials"], :author=>["Leiserson, Charles E and Li, Liyun and Maza, Marc Moreno and Xie, Yuzhen"], :booktitle=>["International Congress on Mathematical Software"], :pages=>["342--353"], :year=>["2010"], :organization=>["Springer"]}]

**Comment by MrMinimal64** [comment]
@whedon check references

**Comment by dpsanders** [comment]
I guess you're fixing the missing DOIs, thanks!

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/MCSE.2011.37 is OK
- 10.1145/2833157.2833162 is OK
- 10.1098/rstl.1819.0023 is OK
- 10.1007/978-3-030-01177-2_5 is OK
- 10.1145/980175.980179 is OK
- 10.1007/s006070070002 is OK
- 10.15807/jorsj.51.29 is OK
- 10.1090/s0025-5718-1990-0993925-1 is OK
- 10.1016/j.cpc.2013.05.008 is OK
- 10.1007/978-3-642-15582-6_55 is OK
- 10.1016/s0377-0427(02)00517-4 is OK

MISSING DOIs

- 10.1201/9781315120195-1 may be a valid DOI for title: Introduction to numerical analysis

INVALID DOIs

- None
```

**Comment by MrMinimal64** [comment]
The last remaining DOI is actually incorrect. Many books have this name.

**Comment by dpsanders** [comment]
It seems to be https://doi.org/10.1017/CBO9780511612916 ?

**Comment by MrMinimal64** [comment]
Indeed. Thanks. I will add it.

**Comment by MrMinimal64** [comment]
The current version number is v2.1.1
and the DOI 10.5281/zenodo.4060365

**Comment by dpsanders** [comment]
@whedon set version as 2.1.1

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by dpsanders** [comment]
@whedon set archive as 10.5281/zenodo.4060365

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by dpsanders** [comment]
@whedon set 2.1.1 as version

**Comment by whedon** [comment]
OK. 2.1.1 is the version.

**Comment by dpsanders** [comment]
@whedon set 10.5281/zenodo.4060365 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4060365" target="_blank">10.5281/zenodo.4060365</a> is the archive.

**Comment by dpsanders** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/MCSE.2011.37 is OK
- 10.1145/2833157.2833162 is OK
- 10.1098/rstl.1819.0023 is OK
- 10.1017/CBO9780511612916 is OK
- 10.1007/978-3-030-01177-2_5 is OK
- 10.1145/980175.980179 is OK
- 10.1007/s006070070002 is OK
- 10.15807/jorsj.51.29 is OK
- 10.1090/s0025-5718-1990-0993925-1 is OK
- 10.1016/j.cpc.2013.05.008 is OK
- 10.1007/978-3-642-15582-6_55 is OK
- 10.1016/s0377-0427(02)00517-4 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1767

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1767, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@MrMinimal64 I'm processing your paper for acceptance in JOSS. I have read your paper and ask you to work on the remaining minor points below: 

- [ ] Make edit the ZENODO archive meta-data and make sure that in terms of title and author list it matches that of the paper

- [ ] The version tag for the archived version should also match the reviewed software version tag. Should we update the version tag of the reviewed software to `v2.1.1-zenodo` (we can do this for you) or should the ZENODO archive version tag be changed to `v2.1.1`? 

- [ ] Please add city and country information to your affiliations

- [ ] Your paper is about to be processed for acceptance. I suggest you also proofread the draft one more time before we move on, and make any minor changes needed. In particular make sure the authors, affiliations, and acknowledgements are accurate/complete. 

Kevin Moerman 
_AEiC JOSS_

**Comment by dpsanders** [comment]
@whedon remove @alex-konovalov as reviewer

**Comment by whedon** [comment]
OK, @alex-konovalov is no longer a reviewer

**Comment by MrMinimal64** [comment]
I uploaded the version `v2.1.1` to zenodo. The difference is just one missing DOI in a reference, but this should not make a difference for the software archive right? I think this version tag is preferable.
The DOI now changed to:
10.5281/zenodo.4062422
I updated the metadata of the software archive and added city and country to my affiliations.
The authors, affiliations, and acknowledgements should now be complete.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set 10.5281/zenodo.4062422 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4062422" target="_blank">10.5281/zenodo.4062422</a> is the archive.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/MCSE.2011.37 is OK
- 10.1145/2833157.2833162 is OK
- 10.1098/rstl.1819.0023 is OK
- 10.1017/CBO9780511612916 is OK
- 10.1007/978-3-030-01177-2_5 is OK
- 10.1145/980175.980179 is OK
- 10.1007/s006070070002 is OK
- 10.15807/jorsj.51.29 is OK
- 10.1090/s0025-5718-1990-0993925-1 is OK
- 10.1016/j.cpc.2013.05.008 is OK
- 10.1007/978-3-642-15582-6_55 is OK
- 10.1016/s0377-0427(02)00517-4 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1774

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1774, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1311968139010281473) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1775
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02392](https://doi.org/10.21105/joss.02392)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02392/status.svg)](https://doi.org/10.21105/joss.02392)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02392">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02392/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02392/status.svg
   :target: https://doi.org/10.21105/joss.02392
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02392/status.svg)](https://doi.org/10.21105/joss.02392)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by Kevin-Mattheus-Moerman** [comment]
Congratulations on your accepted paper @MrMinimal64 ! :tada: 

Thanks for your review efforts @henrik227, @saschatimme!!! 

And thank you @dpsanders for editing this submission!

**Comment by dpsanders** [comment]
Thanks @Kevin-Mattheus-Moerman!

Congratulations @MrMinimal64 on your paper!

And many thanks to @henrik227 and @saschatimme for your hard work on the reviews!

**Comment by MrMinimal64** [comment]
Thank you all for enabling my first sole publication!

One thing I noticed. Since there is an underscore in the title, using the linked BibTex silently breaks a latex document.
I think it is very important to fix this, since the readers otherwise search for errors in their documents forever.
All underscores need to be escaped:

`title = {multivar\_horner: A Python package for computing Horner factorisations of multivariate polynomials},`

**Comment by alex-konovalov** [comment]
@MrMinimal64 congratulations! Thanks to the reviewers - sorry, haven't been checking as active on github recently, missed reminders.
